calculator.cc

```diff
@@ -0,0 +1,84 @@
+#include "calculator.h"
+#include "calculator_exception.h"
+
+#include <array>
+#include <algorithm>
+#include <iterator>
+#include <sstream>
+#include <stack>
+#include <vector>
+
+namespace dev {
+
+IToken::IToken(const std::string& token):token_(token) {}
+
+Operand::Operand(const std::string& token):IToken(token) {
+    try {
+        value_ = std::stod(token);
+    } catch (const std::invalid_argument&) {
+        throw InvalidOperandException();
+    } catch (const std::out_of_range&) {
+        throw OperandOutOfRangeException();
+    }
+}
+
+double Operand::value() const {
+    return value_;
+}
+
+Operator::Operator(const std::string& token):IToken(token) {
+    // ToDo validate operator and raise exception
+    value_ = token[0];
+}
+
+char Operator::value() const {
+    return value_;
+}
+
+bool Calculator::isOperator(const std::string& token) {
+    std::array<std::string, 4> operators = {"+", "-", "*", "/"};
+    return (std::find(operators.begin(), operators.end(), token) != operators.end());
+}
+
+double Calculator::calc(const std::string& expression) {
+    std::istringstream iss(expression);
+    std::vector<std::string> tokens((std::istream_iterator<std::string>(iss)),
+                                     std::istream_iterator<std::string>());
+
+    std::stack<std::string> stkTokens;
+    for (auto token : tokens) {
```

Comment:
Separate validation and calculation

Reply:
for (auto token : tokens) {
 if (operand) 
   push to stack 
}
operators  = extract operators (tokens)
for (operator : operators ) {
  apply operator (pop, pop) -> stack 
}
check lengh of the stack

Reply:
The problem is we cannot just stack up all the operands. It will mess up the order of execution. When we iterate through the tokens, and if we encounter an operator then we have to pop two operands and push it with the result of the operation. So we cannot maintain if we try to do it in two different cycles I suppose(I tried it).

Reply:
Done.